x-pack/plugins/logstash/public/application/pipeline_edit_view.tsx

```diff
@@ -0,0 +1,153 @@
+// @ts-ignore
+import { PipelineEditor } from './components/pipeline_editor';
+// @ts-ignore
+import { Pipeline } from '../models/pipeline';
+import { ManagementAppMountParams } from '../../../../../src/plugins/management/public';
+// @ts-ignore
+import * as Breadcrumbs from './breadcrumbs';
+
+const usePipeline = (
+  pipelineService: any,
+  logstashLicenseService: any,
+  toasts: ToastsStart,
+  shouldClone: boolean,
+  id?: string
+) => {
+  const mounted = usePromise();
+  const [pipeline, setPipeline] = useState<any | null>(null);
+
+  useLayoutEffect(() => {
+    (async () => {
+      if (!id) {
+        return setPipeline(new Pipeline());
+      }
+
+      try {
+        const result = await mounted(pipelineService.loadPipeline(id) as Promise<any>);
+        setPipeline(shouldClone ? result.clone : result);
+      } catch (e) {
+        await logstashLicenseService.checkValidity();
+        if (e.status !== 403) {
+          toasts.addDanger(
+            i18n.translate('xpack.logstash.couldNotLoadPipelineErrorNotification', {
+              defaultMessage: `Couldn't load pipeline. Error: '{errStatusText}'.`,
+              values: {
+                errStatusText: e.statusText,
+              },
+            })
+          );
+        }
+      }
+    })();
+  }, [pipelineService, id, mounted, shouldClone, logstashLicenseService, toasts]);
+
+  return pipeline;
+};
+
+const useIsUpgraded = (upgradeService: any) => {
+  const [isUpgraded, setIsUpgraded] = useState<null | boolean>(null);
+  const mounted = usePromise();
+
+  useLayoutEffect(() => {
+    mounted(upgradeService.executeUpgrade() as Promise<boolean>).then(result =>
+      setIsUpgraded(result)
+    );
+  }, [mounted, upgradeService]);
+
+  return isUpgraded;
+};
+
+interface EditProps {
+  pipelineService: any;
+  logstashLicenseService: any;
+  upgradeService: any;
+  toasts: ToastsStart;
+  history: History;
+  setBreadcrumbs: ManagementAppMountParams['setBreadcrumbs'];
+
+  // URL params
+  id?: string;
+}
+
+export const PipelineEditView: React.FC<EditProps> = ({
+  pipelineService,
+  logstashLicenseService,
+  upgradeService,
+  toasts,
+  history,
+  setBreadcrumbs,
+  id,
+}) => {
+  const params = new URLSearchParams(history.location.search);
+  const shouldRetry = params.get('retry') === 'true';
+  const shouldClone = params.get('clone') === '';
+
+  const pipeline = usePipeline(pipelineService, logstashLicenseService, toasts, shouldClone, id);
+  const isUpgraded = useIsUpgraded(upgradeService);
+
+  const onRetry = useCallback(() => {
+    const newParams = new URLSearchParams(history.location.search);
+    newParams.set('retry', 'true');
+    history.replace({ search: newParams.toString() });
+  }, [history]);
+  const close = useCallback(() => {
+    history.push('/');
+  }, [history]);
+  const open = useCallback(
+    (newId: string) => {
+      history.push(`/pipeline/${newId}/edit`);
+    },
+    [history]
+  );
```

Comment:
NIT: `useCallback` make sense for `onRetry` (even if debatable), but it seems unnecessary here.

Reply:
Correct me if I'm wrong (I'm still new to hooks), but I thought the purpose of useCallback was to avoid creating a new function on each render, possibly triggering unnecessary re-renders of the child components. I'm basically just using this for memoization.

Reply:
Yea you are right. Sorry, I thought `history` was also passed down to the child component (which would have made the useCallback usage useless) but it is not. All good :)